crates/storage/snapshot/difflayer.rs

```diff
@@ -0,0 +1,262 @@
+        }
+
+        let hit = self
+            .diffed
+            .contains_input(BloomInput::Hash(hash.as_fixed_bytes()));
+
+        // If bloom misses we can skip diff layers
+        if !hit {
+            return self.origin.get_account(hash, layers);
+        }
+
+        // Start traversing layers.
+        self.get_account_traverse(hash, layers)
+    }
+
+    pub fn get_storage(
+        &self,
+        account_hash: H256,
+        storage_hash: H256,
+        layers: &Layers,
+    ) -> Result<Option<U256>, SnapshotError> {
+        if self.stale {
+            return Err(SnapshotError::StaleSnapshot);
+        }
+
+        if let Some(value) = self
+            .storage
+            .get(&account_hash)
+            .and_then(|x| x.get(&storage_hash))
+        {
+            return Ok(Some(*value));
+        }
+
+        let bloom_hash = account_hash ^ storage_hash;
+        let hit = self
+            .diffed
+            .contains_input(BloomInput::Hash(bloom_hash.as_fixed_bytes()));
+
+        // If bloom misses we can skip diff layers
+        if !hit {
+            return self.origin.get_storage(account_hash, storage_hash, layers);
+        }
+
+        // Start traversing layers.
+        self.get_storage_traverse(account_hash, storage_hash, layers)
+    }
+
+    pub fn stale(&self) -> bool {
+        self.stale
+    }
+
+    pub fn mark_stale(&mut self) -> bool {
+        let old = self.stale;
+        self.stale = true;
+        old
+    }
+
+    pub fn parent(&self) -> H256 {
+        self.parent
+    }
+
+    pub fn update(
+        &self,
+        block: BlockHash,
+        state_root: H256,
+        accounts: HashMap<H256, Option<AccountState>>,
+        storage: HashMap<H256, HashMap<H256, U256>>,
+    ) -> DiffLayer {
+        let mut layer = DiffLayer::new(
+            self.block_hash,
+            self.origin.clone(),
+            block,
+            state_root,
+            accounts,
+            storage,
+        );
+
+        layer.rebloom(self.origin.clone(), Some(self.diffed));
+
+        layer
+    }
+
+    pub fn origin(&self) -> Arc<DiskLayer> {
+        self.origin.clone()
+    }
+
+    pub fn diffed(&self) -> Bloom {
+        self.diffed
+    }
+
+    // skips bloom checks, used if a higher layer bloom filter is hit
+    pub fn get_account_traverse(
+        &self,
+        hash: H256,
+        layers: &Layers,
+    ) -> Result<Option<AccountState>, SnapshotError> {
+        if self.stale {
+            return Err(SnapshotError::StaleSnapshot);
+        }
+
+        // If it's in this layer, return it.
+        if let Some(value) = self.accounts.get(&hash) {
+            return Ok(value.clone());
+        }
+
+        // delegate to parent
+        match &layers[&self.parent] {
+            Layer::DiskLayer(disk_layer) => disk_layer.get_account(hash, layers),
+            Layer::DiffLayer(diff_layer) => diff_layer
+                .read()
+                .unwrap()
```

Comment:
Could we use `?` instead of unwrapping here?

Reply:
fixed